Aqui empieza el curso de ML aplicado

In [84]:
import datetime
print(f'The last time is:  {datetime.datetime.now()}')

The last time is:  2026-04-07 14:40:59.328629


In [85]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import sklearn

Bueno, lo mas importante a la hora de estudiar ML y los modelos es entender los pasos de lo que estamos haciendo, entender la teoria, identificar el problema, tener los datos libros. 
Si tenemos un problema de clasificacion, un problema de prediccion o un problema de agrupacion. Los pasos que tenemos que considerar para el workflow de Scikit-learn son: 
1. Getting Data (find columns, labels, prepair train and test steps)
2. Choosing a model for our problem 
3. Ajustar el modelo a la data que tenemos y usarlo para obtener la predicción
4. Evaluar el modelo 
5. Experimentar para mejorar
6. Guardar el modelo para que alguien mas lo pueda usar, o yo mismo en el futuro. 

Para el primer ejemplo vamos a usar un dataset llamado heart-disease.csv

In [86]:
url = 'https://raw.githubusercontent.com/Juan-Vega18/ML_full_course/main/my_notebooks/data/heart-disease.csv'
df = pd.read_csv(url)

In [87]:
#Ahora vamos a ver algunas propiedades de nuestro dataset
df.info()
df.shape
df.head(5)
#df['target'].unique()
#df['restecg'].unique()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1


Encontramos un conjunto de 14 columnas con 303 filas. La fase de entender como y que especificamente compone nuestro dataset es absolutamente clave. Tenemos target que a pesar de que parece ser una variable numerica en realidad es booleana, el dataset esta completisimo, no tiene vacios etc. 

En el casode este proyecto, esa variable booleana es precisamente la que queremos predecir, si está o no enfermo del corazón, 1 significa si, 0 significa no. Las demas columnas se llamaran 'features' y las usaremos para intentar predecir a target. 

In [88]:
x = df.drop('target', axis=1) #Creamos los x de los parametros. eliminamos la columna target
y = df['target'] #Creamos la y que será nuestra predicción 
x.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2


In [89]:
#Ahora vamos a revisar los valores de y que tenemos 
print(f'Los valores de la variable estan distribuidos de la siguiente forma: {y.value_counts()}')
y.unique()

Los valores de la variable estan distribuidos de la siguiente forma: target
1    165
0    138
Name: count, dtype: int64


array([1, 0])

Ahora, algo que es absolutamente importante en el ML es entender que la data debe ser divida en dos grupos, training set and test set. Es decir, un grupo de datos se usará para entrenar el modelo, el otro para probarlo. Usualmente se combina en proporciones como 80/20, 70/30, 75/25, tambien es posible tener el validation set que es un dataset extra que se usa durante el entrenamiento pero por ahora no lo vamos a usar. 

Scikit-learn ya incluye sklearn.model_selection.train_test_split que es la funcion de la libreria para dividir el dataset en training sets. 


In [90]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(x,y,test_size=0.25)
print(X_train.shape)
print(type(X_train))
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

#Aqui ya dividimos la data, notemos que la funcion train_test_split asigna a 4 variables los 
#datos correspondientes y los subdivide en pandas.core.frame.DataFrame luego veremos que significa
# y que tiene de diferente con el pd.Dataframe tipico.

X_train

(227, 13)
<class 'pandas.core.frame.DataFrame'>
(76, 13)
(227,)
(76,)


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
287,57,1,1,154,232,0,0,164,0,0.0,2,1,2
282,59,1,2,126,218,1,1,134,0,2.2,1,1,1
197,67,1,0,125,254,1,1,163,0,0.2,1,2,3
158,58,1,1,125,220,0,1,144,0,0.4,1,4,3
164,38,1,2,138,175,0,1,173,0,0.0,2,4,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...
188,50,1,2,140,233,0,1,163,0,0.6,1,1,3
71,51,1,2,94,227,0,1,154,1,0.0,2,1,3
106,69,1,3,160,234,1,0,131,0,0.1,1,1,2
270,46,1,0,120,249,0,0,144,0,0.8,2,0,3


Ahora que tenemos los datos divididos es necesario escoger el modelo y los hiperparametros. Escoger el modelo depende normalmente de cual sea el problema que estamos trabajando, por ejemplo, Scikit-learn tiene recomendaciones si el problema es de clasificacion o predicción. 
En la web de scikit-learn tenemos un mapa muy util que explica que modelo usar, cuando y bajo que condiciones. 
La primera y mas importante es la cantidad de ejemplos (se recomienda mas de 50). 

Tenemos problemas de clasificacion, regresion, clustering y reduccion de la dimension. 
Aquí es donde el estudiante (yo) debería aprender bien y antes de que el curso acabe haber probado todos o la gran mayoría de los algoritmos, la programación no necesariamente es lo mas importante, lo que importa aquí es saber bien que modelo usar, cuando, y por qué. 

Tambien tenemos los hiperparametros del modelo, que si sabemos matematicamente lo que estamos haciendo, podemos tocarlos y revisar si el modelo mejora o no. Se van ajustando. Aquí tambien es muy importante que el estudiante entienda que está pasando en el interior de la maquina, de nada sirve tener el modelo, el codigo, resultados si no soy capaz de modificar parametros (no entiendo nada).

En el caso de este proyecto, estamos trabajando con un clasificador, booleano. Si o no. Asi que el curso nos recomienda usar empezar con el classifier Random forest.

In [91]:
from sklearn.ensemble import RandomForestClassifier #Aqui hay muchos interesantes 

clf = RandomForestClassifier() #Aqui inicializamos el modelo, y vamos a ver los parametros
#que podemos modificar 
clf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'sqrt',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'monotonic_cst': None,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

#Por ahora, 
no vamos a tocarlos, los dejaremos para mas adelante. Usaremos entonces
#los parametros por defecto de Sckit-learn. 

Ahora, vamos a Fit the model to the data and use it to make a prediction. Es lo que queremos. Si tenemos etiquetas (aprendizaje supervisado) el modelo trabaja en encontrar la relacion entre los datos y las etiquetas. En el caso de que no hayan etiquetas, el modelo intenta encontrar patrones y agrupar ejemplos similares juntos. 
La mayoria de los modelos de Scikit-Learn tienen el methodo fit(X,Y) integrado, donde X son las caracteristicas (features) e y son las etiquetas. 
En nuestro caso, vamos a empezar ajustando un modelo a los datos de entrenamiento separados. 



In [92]:

clf.fit(X=X_train, y=y_train) #clf es nuestro modelo, y a ese modelo aplicamos parametros

RandomForestClassifier()

Ahora, ya que tenemos el "modelo ajustado" viene la parte central de esto. intentar hacer una predicción. Despues de que el modelo ya está entrenado, ahora lo que hay que hacer es predecir, usando predict(). En otras palabras, usar el modelo con data nueva, que no haya visto y que no esté etiquetada. (Ojo, aqui es importente), la data que daremos para predecir debe tener el mismo shape que la data que usamos para entrenar (matematicamente tiene todo el sentido pero igual se aclara). 

Para eso, usaremos la data de test

In [93]:
y_preds = clf.predict(X=X_test)
print(y_preds)

#Y si queremos analizar una unica fila, dado que es un 
#dataset 2D esperamos esa entrada

predic_1 = clf.predict(X_test.iloc[[0]])
predic_1

[0 1 0 0 1 1 1 0 0 0 1 0 1 0 1 1 1 0 0 0 1 0 1 1 1 1 1 1 0 1 0 0 0 0 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 0 0 0 0 1 1 0 0 0 1 0 0 0 1 0 0 1 0 0 1 1 1 1 1 1 1
 0 1]


array([0])

Ahí tenemos el array de predicciones, ahora lo que nos interesa es evaluar el modelo. Cada modelo estimador tiene un metodo integrado llamada score(). Este metodo compara que tan bien le fue al modelo aprendiendo los patrones entre las caracteristicas y las etiquetas. El metodo score() usa una metrica estandar de evaluacion para medir los resultados del modelo. 

En el caso de un clasificador (nuestro modelo), una de las metricas de evaluacion mas comunes es 'accuracy', que es la fraccion de predicciones correctas sobre el total de predicciones. 

In [94]:
#Evaluate the model on the training set

train_acc = clf.score(X=X_test, y=y_test)
print(f'The models accuracy on the training dataset is: {train_acc*100} %')

The models accuracy on the training dataset is: 82.89473684210526 %


Existen algunos otros metodos de evaluacion para nuestro modelo, estos vienen del modulo sklearn.metrics y tenemos: 
Classification report (y true, y true)
Construye un reporte mostrando varias metricas de clasificacion, como precision, recall, and F1-score. (luego las entenderemos)

Confusion matrix(y true, y pred). Crea una matriz de confusion para comparar la predición de las etiquetas verdaderas

Accuracy score (y_true, y_pred). Encuentra el puntaje de precisión, esta es la metrica por defecto de los clasificadores. 

In [95]:
#Classification report 
from sklearn.metrics import classification_report

print(classification_report(y_test, y_preds))

              precision    recall  f1-score   support

           0       0.82      0.80      0.81        35
           1       0.83      0.85      0.84        41

    accuracy                           0.83        76
   macro avg       0.83      0.83      0.83        76
weighted avg       0.83      0.83      0.83        76



In [96]:
#Confusion matrix
from sklearn.metrics import confusion_matrix

print(confusion_matrix(y_test, y_preds))

[[28  7]
 [ 6 35]]


In [97]:
#Compute the accuracy score

from sklearn.metrics import accuracy_score

accuracy_score(y_test, y_preds)

0.8289473684210527

Despues de todo esto que vimos, nos damos cuenta que en realidad la implementacion es muy sencilla, y obtuvimos de hecho un 'buen rendimiento', supongo, dado que en el futuro de este curso vamos a estudiar como saber si el rendimiento fue bueno o no etc. 

El primer modelo es usualmente tomado como baseline, basicamente lo usamos para revisar la predicción base y es un indicador de que debemos hacer en el futuro, aqui debe aparece un criterio importante y es ir aprendiendo cuando con el baseline vale la pena continuar con un estimador, o aprender que parametros parece mas importante modificar etc. Importante hacer la prueba e ir escalando. 

El siguiente paso es en el workflow es intentar mejorar nuestro modelo base (que en teoria ya deberia tener un margen decente para empezar)

Como? 
Bueno, una de las cosas mas importantes en el Machine Learning es experimentar, experimentar y experimentar, es la forma correcta. Y hay dos caminos que hoy podemos mencionar para los pasos a continuacion: 

1. Experimentos enfocados en el modelo
2. Experimentos enfocados en los datos

Los experimentos enfocados en el modelo pueden incluir modificar parametros mucho mas complejos o usar un modelo mas complicado, 

La perspectiva enfocada en los datos puede incluir obtener mas datos o de mejor calidad, eso podria hacer que el modelo que ya tenemos tenga mas chances de aprender mejor los patrones. 

#QUE ES LA CROSS-VALIDATION
Es ua forma de asegurar que los resultados que se estan obteniendo son consistentes a lo largo de los intentos y pruebas de entrenamiento, usar multiples versiones de entrenamiento y sets de datos influye, en lugar de solo suerte por el orden de el set de datos de entrenamiento y de pruebas que se crearon. 
Hay dos elementos importantes: 
1. Probar diferentes hiperparámetros 
2. Todos los diferentes parametros deben ser cross-validated. (Es importante recordar que si tenemos series de tiempo no se deben mezclar) 

En este caso, nuestro modelo RandomForestClassifier(), vamos a empezar probando diferentes valores de n_stimators que basicamente es la cantidad de arbol en el Rando Forest. 
Por defecto tenemos n_estimators = 100, entonces, vamos a intentar valores entre 100 y 200 y veremos que sucede.. ¿Generalmente es mejor? 


In [98]:
np.random.seed(42)
for i in range(100,200,10):
    print(f'Trying model with {i} estimators... ')
    model = RandomForestClassifier(n_estimators=i).fit(X_train, y_train)
    print(f'Model accuray on test set: {model.score(X_test, y_test)}')
    print("")

Trying model with 100 estimators... 
Model accuray on test set: 0.8157894736842105

Trying model with 110 estimators... 
Model accuray on test set: 0.8289473684210527

Trying model with 120 estimators... 
Model accuray on test set: 0.8289473684210527

Trying model with 130 estimators... 
Model accuray on test set: 0.8289473684210527

Trying model with 140 estimators... 
Model accuray on test set: 0.8421052631578947

Trying model with 150 estimators... 
Model accuray on test set: 0.8421052631578947

Trying model with 160 estimators... 
Model accuray on test set: 0.8552631578947368

Trying model with 170 estimators... 
Model accuray on test set: 0.8552631578947368

Trying model with 180 estimators... 
Model accuray on test set: 0.8421052631578947

Trying model with 190 estimators... 
Model accuray on test set: 0.8421052631578947



Notamos que cuando usamos diferente numero de estimadores obtenemos diferentes puntajes. En el caso de las pruebas de 10 en 10 encontramos que con 110, 120, 150, 170 encontramos que tenemos el mismo score. Por lo tanto, no es cuestión aumentar los estimadores arbitrariamente. 

Usaremos sklearn.model selection.cross val score para medir el resultado con 5 diferentes sets de entrenamiento y prueba. (train and test sets)

Podemos medir esto configurando: 
cross_val_score(X, y, cv=5)
Ddonde X el dataset completo, y es el full set de etiquetas (basicamente el dato predictivo) y cv es el numero de sets de entrenamiento y test splits cross_val_score creará automaticamente de nuestra data (en nuetro caso) 5 diferentes divisiones, esto es conocido como 5-fold cross-validation. 

In [125]:
from sklearn.model_selection import cross_val_score

#With cross validation 
np.random.seed(42) #Elegimos una semilla para que el calculo sea rep1licable. 
for i in range(100,200,10):
    print(f'The model with {i} estimators...')
    model = RandomForestClassifier(n_estimators=i)

    scores = cross_val_score(model, X_train, y_train, cv=5)
    print(f'Scores: {scores}')
    print(f'Mean accuracy: {scores.mean()*100:.5f}%')
    print("")

The model with 100 estimators...
Scores: [0.82608696 0.76086957 0.73333333 0.86666667 0.82222222]
Mean accuracy: 80.18357%

The model with 110 estimators...
Scores: [0.82608696 0.86956522 0.77777778 0.84444444 0.82222222]
Mean accuracy: 82.80193%

The model with 120 estimators...
Scores: [0.84782609 0.84782609 0.75555556 0.84444444 0.75555556]
Mean accuracy: 81.02415%

The model with 130 estimators...
Scores: [0.84782609 0.82608696 0.75555556 0.86666667 0.8       ]
Mean accuracy: 81.92271%

The model with 140 estimators...
Scores: [0.84782609 0.80434783 0.77777778 0.86666667 0.8       ]
Mean accuracy: 81.93237%

The model with 150 estimators...
Scores: [0.80434783 0.84782609 0.8        0.82222222 0.75555556]
Mean accuracy: 80.59903%

The model with 160 estimators...
Scores: [0.82608696 0.80434783 0.73333333 0.84444444 0.8       ]
Mean accuracy: 80.16425%

The model with 170 estimators...
Scores: [0.82608696 0.86956522 0.73333333 0.84444444 0.8       ]
Mean accuracy: 81.46860%

The mode

¿Que modelo tiene la mejor cross-validation score? Este es usualmente un mejor indicador de la calidad del modelo que un unico puntaje de precisión. 
En lugar de hacer el set up y seguir los resultados de los experimentos manualmente, podemos hacer que Scikit-Lean haga la exploracion por nosotros. 

Scikit-Learn's sklearn.model selection.GridSearch CV es una forma de de checar diferentes hiperparametros y valores y evaluar automaticamente cual rinde mejor. 

In [103]:
#Another way to do it with GridSearchCV...
np.random.seed(42)
from sklearn.model_selection import GridSearchCV

#Definimos los parametros para probar y los metemos en un diccionario
# estos pueden ser cualquiera de los hiperparametros del modelo 
param_grid = {'n_estimators': [i for i in range(100,200,10)]}

#Setup de grid search 

grid = GridSearchCV(estimator=RandomForestClassifier(), 
                    param_grid=param_grid,
                    cv=5,
                    verbose=1)

#Fit the grid search to the data
grid.fit(x,y) #Metemos la data completa

#Find the best parameters: 

print(f'The best parameter values are: {grid.best_params_}')
print(f'With a score of: {grid.best_score_*100:.2f}%')


Fitting 5 folds for each of 10 candidates, totalling 50 fits
The best parameter values are: {'n_estimators': 120}
With a score of: 82.82%


Y aqui podemos fijarle al modelo el mejor estimador

In [104]:
clf = grid.best_estimator_
clf

RandomForestClassifier(n_estimators=120)

Ahora que ya hemos cross-valited el modelo, podemos ajustar y medir el puntaje en nuestro original single train/test split de la data. 

In [118]:
np.random.seed(42)
clf = clf.fit(X_train, y_train)

# Find the best model scores on our single test split
# esto puede ser menor que el puntaje cross-validated dado que es solo una
#division de la data. 

print(f'Best model score on single split of the data: {clf.score(X_test, y_test)*100:.2f}%')

Best model score on single split of the data: 81.58%


Aquí noto algo RARISIMO y es que cada vez que ajustamos el modelo SIN una semilla obtenemos scores diferentes, esto es bien interesante dado que la data se ajusta de diferentes maneras, lo que es bien raro... Si fijo un np.random.seed(42) mantengo el mismo rendimiento... pero si la data YA está separada... que hace que el resultado sea tan distinto? vale la pena evaluarlo. 

Ahora lo que queremos hacer guardar el modelo para que alguien mas pueda usarlo o para usarlo mas tarde. 
Usamos el modelo pickle  

In [124]:
import pickle 

#Save an existing model to file 
pickle.dump(model, open("models/random_forest_model_1.pkl", "wb"))

Me gustaria tomarme unos parrafos para explicar realmente que el cross-validation ya que creo que es muy importante tenerlo bien claro. Nosotros tenemos un set de data completo D. Primero se divide en 2 subconjuntos D_{train} (\approx80%) y D_{test} (\approx 20%). El conjunto D_{test} se reserva completamente y no se toca durante el entrenamiento ni la selección del modelo. 

Sobre D_{train} se aplica k-fold cross validation, entonces, que se hace:
1. Se divide D_{train} en k subconjuntos que llamaremos folds, todos del mismo tamaño 
2. Para cada iteracion segun el numero de k folds: 
2.1 Se entrena el modelo con k-1 folds 
2.2 Se evalua con con el fold restante. 
Esto singifica basicamente. tenemos FOLD1, FOLD2, FOLD3, FOLD4, FOLD5
Primer entrenamiento: TRAIN = FOLD1,FOLD3,FOLD4,FOLD5. TEST = FOLD2 y así hasta que se hayan bajarado todos. 
3. Se obtienen k metricas, como accuracy o score. 
4. Se calcula el promedo de las k metricas. 

Y este promedio ESTIMA el rendimiento esperado del modelo sobre datos no vistos. 
Y aquí hay algo clave y brutalmente importante. CROSS-VALIDATION NO GARANTIZA QUE EL TEST SEA REPRESENTATIVO. 

Simplemente nos sirve para reducir la incertidumbre en la estimación del modelo antes de evaluarlo en el test final. 

Y Ahora, ya que dominamos cross-validation (al menos de forma teórica) tenemos que pasar con esta idea base: 
'Si ya entendí que un modelo no debe evaluarse con una sola partición, entonces tampoco debería elegir sus hiperparámetros "a ojo" con una sola partición'. 
Y eso es lo que hace GridSearchCV

Con nuestro ejemplo hicimos Random forest cambiando los estimadores en un rango: 

In [126]:
for i in range(100,200,10):
    ...

Estamos variando los n_estimators (es solo uno de los hiperparametros que tenemos disponibles) y luego comprobabamos los scores. Eso se hace a mano por el ejemplo, pero GridSearchCV lo hace por nosotros. 

Entonces, GridSearchCV toma: 
1. El modelo 
2. una rejilla de hiperparámetros
3. Una estrategia de cross-validation 
Entonces, prueba todas las combinaciones de hiperparametros (potente)
para cada combinación aplica crossvalidation 
calcula el score promedio
Luego escoje la mejor combinación 
y luego entrega un modelo final con esa mejor combinación jajajaj. Básicamente una sola función de librería resuelve un montón de problemas. 
Un grid (como vimos ya en el ejemplo) se ve así: 

In [127]:
param_grid = {
    "n_estimators": [100,200,300],
    "max_depth": [None, 5, 10],
    "min_samples_split":[2,4]
    }

Entonces, tenemos 3 opciones para n_estimators, 3 para max_depth, 2 para min_samples_split, entonces, tenemos 3x3x6 = 18 configuraciones principales, y si usamos cross-validation con k=5, entonces: 18x5 = 90 entrenamientos. 
Entonces. ahi nos damos cuenta de la potencia de esa función. Aquí un ejemplo: 

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV, train_test_split

# 1. Separar train y test
X_train, X_test, y_train, y_test = train_test_split(
    x, y, test_size=0.2, random_state=42 
)

# 2. Definir modelo base
model = RandomForestClassifier(random_state=42)

# 3. Definir grid de hiperparámetros
param_grid = {
    "n_estimators": [100, 200, 300],
    "max_depth": [None, 5, 10],
    "min_samples_split": [2, 4, 6]
}

# 4. Grid search con cross-validation
grid = GridSearchCV(
    estimator=model,
    param_grid=param_grid,
    cv=5,
    scoring="accuracy", #La metrica que queremos maximizar
    n_jobs=-1 #Usamos todos los núcleos del procesador
)

# 5. Ajustar sobre train
grid.fit(X_train, y_train)

# 6. Mejor combinación encontrada
print("Best params:", grid.best_params_)
print("Best CV score:", grid.best_score_)

# 7. Evaluación final en test
best_model = grid.best_estimator_
test_score = best_model.score(X_test, y_test)
print("Test score:", test_score)

Best params: {'max_depth': 5, 'min_samples_split': 2, 'n_estimators': 200}
Best CV score: 0.8098639455782314
Test score: 0.8688524590163934
